app/models/ckb_sync/node_data_processor.rb

```diff
@@ -0,0 +1,325 @@
+    end
+
+    def uncle_block_hashes(node_block_uncles)
+      node_block_uncles.map { |uncle| uncle.header.hash }
+    end
+
+    def generate_address_in_advance(cellbase)
+      return if cellbase.witnesses.blank?
+
+      lock_script = CkbUtils.generate_lock_script_from_cellbase(cellbase)
+      address = Address.find_or_create_address(lock_script)
+      LockScript.find_or_create_by(
+        args: lock_script.args,
+        code_hash: lock_script.code_hash,
+        address: address
+      )
+    end
+
+    def build_block(node_block)
+      header = node_block.header
+      epoch_info = CkbUtils.get_epoch_info(header.epoch)
+      cellbase = node_block.transactions.first
+
+      generate_address_in_advance(cellbase)
+
+      Block.new(
+        difficulty: header.difficulty,
+        block_hash: header.hash,
+        number: header.number,
+        parent_hash: header.parent_hash,
+        seal: header.seal,
+        timestamp: header.timestamp,
+        transactions_root: header.transactions_root,
+        proposals_hash: header.proposals_hash,
+        uncles_count: header.uncles_count,
+        uncles_hash: header.uncles_hash,
+        uncle_block_hashes: uncle_block_hashes(node_block.uncles),
+        version: header.version,
+        proposals: node_block.proposals,
+        proposals_count: node_block.proposals.count,
+        cell_consumed: CkbUtils.block_cell_consumed(node_block.transactions),
+        total_cell_capacity: CkbUtils.total_cell_capacity(node_block.transactions),
+        miner_hash: CkbUtils.miner_hash(cellbase),
+        miner_lock_hash: CkbUtils.miner_lock_hash(cellbase),
+        status: "accepted",
+        reward: CkbUtils.block_reward(node_block),
+        reward_status: header.number.to_i == 0 ? "issued" : "pending",
+        total_transaction_fee: 0,
+        witnesses_root: header.witnesses_root,
+        epoch: header.epoch,
+        start_number: epoch_info.start_number,
+        length: epoch_info.length,
+        dao: header.dao
+      )
+    end
+
+    def build_uncle_block(uncle_block, local_block)
+      header = uncle_block.header
+      local_block.uncle_blocks.build(
+        difficulty: header.difficulty,
+        block_hash: header.hash,
+        number: header.number,
+        parent_hash: header.parent_hash,
+        seal: header.seal,
+        timestamp: header.timestamp,
+        transactions_root: header.transactions_root,
+        proposals_hash: header.proposals_hash,
+        uncles_count: header.uncles_count,
+        uncles_hash: header.uncles_hash,
+        version: header.version,
+        proposals: uncle_block.proposals,
+        proposals_count: uncle_block.proposals.count,
+        witnesses_root: header.witnesses_root,
+        epoch: header.epoch,
+        dao: header.dao
+      )
+    end
+
+    def build_ckb_transactions(local_block, transactions)
+      transaction_index = 0
+      transactions.map do |transaction|
```

Comment:
```suggestion
      transactions.each_with_index.map do |transaction, transaction_index|
```

Reply:
https://gist.github.com/spider-network/8b0e168a932731eade6f `each` may faster than `each_with_index`

Reply:
https://media.pragprog.com/titles/adrpo/iterators.pdf this show me  `each_with_index` will create  additional T_NODE. so I choose to use each.